internal/common/utils/pods.go

```diff
@@ -20,17 +20,47 @@ import (
 	"github.com/palantir/k8s-spark-scheduler/internal/common"
 	"github.com/palantir/witchcraft-go-logging/wlog/svclog/svc1log"
 	v1 "k8s.io/api/core/v1"
+	clientcache "k8s.io/client-go/tools/cache"
 )
 
 // IsSparkSchedulerPod returns whether the passed object is a spark application pod which has this scheduler in the scheduler spec
 func IsSparkSchedulerPod(obj interface{}) bool {
-	if pod, ok := obj.(*v1.Pod); ok {
-		_, labelFound := pod.Labels[common.SparkRoleLabel]
+	_, isSparkSchedulerPod := getRoleIfSparkSchedulerPod(obj)
+	return isSparkSchedulerPod
+}
+
+// IsSparkSchedulerExecutorPod returns whether the passed object is a spark application pod which has this scheduler in the scheduler spec and is an executor
+func IsSparkSchedulerExecutorPod(obj interface{}) bool {
+	if role, isSparkSchedulerPod := getRoleIfSparkSchedulerPod(obj); isSparkSchedulerPod {
+		return role == common.Executor
+	}
+	return false
+}
+
+// GetPodFromObjectOrTombstone tries to cast the passed object to a Pod object, and if that's not possible tries to get the Pod object from the tombstone
+func GetPodFromObjectOrTombstone(obj interface{}) (*v1.Pod, bool) {
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		tombstone, ok := obj.(clientcache.DeletedFinalStateUnknown)
+		if !ok {
+			return nil, false
+		}
+		pod, ok = tombstone.Obj.(*v1.Pod)
+		if !ok {
+			return nil, false
+		}
+	}
+	return pod, true
+}
+
+func getRoleIfSparkSchedulerPod(obj interface{}) (string, bool) {
+	if pod, ok := GetPodFromObjectOrTombstone(obj); ok {
```

Comment:
I think we should just try to cast this to a pod object and not try to extract it from a tombstone.
We use `IsSparkSchedulerPod` and `IsSparkSchedulerExecutorPod` in `FilteringResourceEventHandlers` and now we will return true for dead pods as well, but in order for the `DeleteFunc` to be called on pod deletions, I believe the filter should return false as soon as the pod is deleted